internal/dynamic/plugin/subprocess/subprocess.go

```diff
@@ -0,0 +1,211 @@
+package subprocess
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+// ErrProcessAlreadyStarted is returned when trying to start a subprocess that is already running.
+var ErrProcessAlreadyStarted = errors.New("subprocess already started")
+
+// SubProcess defines the interface for managing a subprocess.
+type SubProcess interface {
+	// Start launches the subprocess and begins monitoring. It returns an error
+	// if the process fails to start initially.
+	Start() error
+	// IsRunning checks if the subprocess is currently running.
+	// Close MUST be called even if IsRunning returns false.
+	IsRunning() bool
+	// Close attempts to gracefully shut down the subprocess. It waits for the process
+	// to exit and cleans up resources.
+	Close(ctx context.Context) error
+}
+
+// Option is a function that can configure a SubProcess.
+type Option func(*subprocess)
+
+// WithLogger allows providing a custom logger for internal library messages.
+func WithLogger(logger *service.Logger) Option {
+	return func(s *subprocess) {
+		s.logger = logger
+	}
+}
+
+// WithStdoutHook allows providing a custom logger for stdout messages.
+func WithStdoutHook(hook func(line string)) Option {
+	return func(s *subprocess) {
+		s.stdoutHook = hook
+	}
+}
+
+// WithStderrHook allows providing a custom logger for stderr messages.
+func WithStderrHook(hook func(line string)) Option {
+	return func(s *subprocess) {
+		s.stdoutHook = hook
+	}
+}
+
+type subprocess struct {
+	cmdArgs    []string
+	env        map[string]string
+	stdoutHook func(line string)
+	stderrHook func(line string)
+	logger     *service.Logger
+
+	cmd    *exec.Cmd
+	mu     sync.Mutex
```

Comment:
It seems we could be using https://pkg.go.dev/sync#OnceValues

Reply:
I don't think so because we need to be able to restart the process